ENH: Rearrange navigation for new leactures, add filtering

DrDrij opened the pull request:
In progress.

- Adjust primary navigation order, add lectures
- Add 6 new lectures to lecture page
- Change layout to grid
- Add filtering
- Add new metadata field for topics

mmcky commented:
@DrDrij I will setup `netlify` previews today if that would help.

DrDrij commented:
Thanks Matt. Ace. I’ll finish the tasks and address any feedback for your
tomorrow.

On Mon, 4 Mar 2024 at 11:24 pm, Matt McKay ***@***.***> wrote:

> @DrDrij <https://github.com/DrDrij> I will setup netlify previews today
> if that would help.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuantEcon/website/pull/121#issuecomment-1977573057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABM5AXOY2Z5TWQS3PEEPFKTYWTYBXAVCNFSM6AAAAABEFW6WQSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSNZXGU3TGMBVG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mmcky commented:
@DrDrij just to confirm the netlify preview mechanism is activated -- but it is part of the checks (as this project is built with jekyll -- and netlify does the building rather than github actions)

Here is the latest link: https://deploy-preview-121--sleepy-aryabhata-461b8b.netlify.app/

<img width="1680" alt="Screenshot 2024-03-05 at 12 29 36 pm" src="https://github.com/QuantEcon/website/assets/8263752/7b5e52f2-36c6-4d60-aedb-9a296dbe1ea7">

and the lecture page
<img width="1680" alt="Screenshot 2024-03-05 at 12 30 12 pm" src="https://github.com/QuantEcon/website/assets/8263752/5d2f9ea0-f24e-4517-bbca-3c07e65cfb91">

mmcky commented:
@jstac when you're up to it would appreciate your thoughts on what dimensions to best filter to find the lectures of interest. 

I think `topic` is the dimension with the most variance. 

While programming language is a natural choice -- it is really one `1 x julia` vs `8 x python` so not the best dimension. 

Maybe we could do

- Introductory (`lecture-python-programming`, `lecture-intro`)
- Economics (`lecture-intro`, `lecture-jax`, `lecture-dynamics`, `lecture-dle`, `lecture-eqm` ...)
- Statistics (`lecture-stats`) # only one for now so maybe not so useful and roll into tools and techniques
- Tools & Techniques (`lecture-python-programming`, `lecture-stats`

jstac commented:
That looks like a pretty good split @mmcky 

How about this similar split?

* Introductory (lecture-python-programming, lecture-intro)
* Programming (lecture-python-programming, jax, julia)
* Topics (various topics lectures)
* Statistics (lecture-stats + data science)

DrDrij commented:
Thanks @jstac and @mmcky. I'll rearrange the metadata fields to match the new topics and we can adjust from there.

Kate had a suggestion of a little "New" badge or dog-ear on the lectures series recently released. I'll pop in an example and we can hide it if it's not really useful for the audience.

DrDrij commented:
Yeah so here is the new mockup — pushed changes to the PR branch.

https://deploy-preview-121--grand-swan-ca5201.netlify.app/lectures/

I had a little fun with it and added the topic filtering you two suggested as well as a "ski slope difficulty indicator icon" and the "new" badges.

**Question: how would you like the page to be ordered? eg. Alphabetically or..?**

Looking forward to feedback.

![lectures-difficulty-topic-filtering1](https://github.com/QuantEcon/website/assets/5886045/f80ad7af-ea9e-4146-85c5-da5534732cb7)

mmcky commented:
thanks @DrDrij now I want to go skiing :-)

@DrDrij I was wondering if when `All` is selected what would it look like to we could have `subheadings` and them sorted into the categories

````
Introductory
-------------
<boxes in this category>

Topics
------
````

and then when a filter is selected the subheadings would just disappear and show the lectures in that filter?

![image](https://github.com/QuantEcon/website/assets/8263752/70f0affb-735d-4925-8a08-b83ff6d841b9)

mmcky commented:
@DrDrij I like the `new` -- are they programmed to disappear after say 6 months of the publish date?

jstac commented:
Nice :-)

Similar to @mmcky 's comment, but maybe a bit easier, perhaps we could order the lectures by difficulty, and then by alphabetical.

So the "All" display shows the green ones first, alphabetically ordered, the blue ones next, alphabetically ordered, and the black ones last, alphabetically ordered.

Then we wouldn't really need the subheadings because all the green ones would be at the top, etc.

mmcky commented:
> Nice :-)
> 
> Similar to @mmcky 's comment, but maybe a bit easier, perhaps we could order the lectures by difficulty, and then by alphabetical.
> 
> So the "All" display shows the green ones first, alphabetically ordered, the blue ones next, alphabetically ordered, and the black ones last, alphabetically ordered.
> 
> Then we wouldn't really need the subheadings because all the green ones would be at the top, etc.

Nice Idea @jstac

DrDrij commented:
How's this @mmcky @jstac ? Did interpret your comments correctly?


![FireShot Capture 061 - Lectures – QuantEcon - 127 0 0 1](https://github.com/QuantEcon/website/assets/5886045/1d4d8f42-4242-4818-ba4f-5e9b810a3c4c)

DrDrij commented:
That looks better.
![image](https://github.com/QuantEcon/website/assets/5886045/9fed4447-72ed-4a37-bc7d-a501f52cb1cd)

mmcky commented:
thanks @DrDrij 

What @jstac and I proposed is to have groups of lectures in the following categories (note: where some lecture series belong to one or more groups). Here the categories are aligned with overall motivation of the series

**Group 1:** `Introductory` (lecture-python-programming, lecture-intro) - group colour `green` (or friendly colour)
**Group 2:** `Programming` (lecture-python-programming, jax, Julia)
**Group 3:** `Topics` (various topics lectures)
**Group 4:** `Statistics` (lecture-stats + data science)

@jstac what do you think about `Topics` being called `Economics`?

jstac commented:
Thanks @DrDrij @mmcky 

My idea of "Topics" is that it covers the specialized topics courses that you do in grad studies after you do the core coursework.

We could change it to "Advanced Topics"?

mmcky commented:
> Thanks @DrDrij @mmcky
> 
> My idea of "Topics" is that it covers the specialized topics courses that you do in grad studies after you do the core coursework.
> 
> We could change it to "Advanced Topics"?

thanks @jstac I agree -- that makes a lot of sense. I think either works. Perhaps let's keep it as `Topics` and we could add a difficulty level indicator if needed. We may have other topics in the future that aren't `Advanced`

DrDrij commented:
Ok folks, I think I get it.

Everything is grouped into difficulty levels and then listed alphabetically. The full set will have green, then blue, and then black icons.

The row of filters up top are for topics, seperate from either grouping above. Selecting a filter will show a subset of lectures that have had that topic assigned to them. Each lecture will have a minimum of 1 topic, but can have multiple.

mmcky commented:
thanks @DrDrij let me know if you need any more specific definitions of the groups. Looking forward to seeing this updated.

DrDrij commented:
@mmcky I think this is ready to go. @jstac Does that look like what you have in mind?

https://deploy-preview-121--grand-swan-ca5201.netlify.app/lectures/#filter=1
(Please clear cache)

<img width="1370" alt="image" src="https://github.com/QuantEcon/website/assets/5886045/7a764a7c-9c45-4866-b489-b6d685d7ee88">

mmcky commented:
thanks @DrDrij nice work! 

Just a few quick questions:

- [ ] The structure is looking good - can I help with by populating the lecture descriptions so we can preview it with Tom?
- [ ] can the `new:` key use `new: {{ true / false }}`? I would find that more intuitive then 1


## Future Idea: 

- [ ] would it be possible to add a `published: {{ date }}` field embedded somewhere on the sites themselves and show an `updated` tag for lectures that have been updated within the last month (or so).

mmcky commented:
thanks @DrDrij I was able to update the metadata for each lecture to remove the `lorem ipsom`

mmcky commented:
@jstac @DrDrij 

latest preview is available here

https://deploy-preview-121--sleepy-aryabhata-461b8b.netlify.app/lectures/#filter=*

mmcky commented:
> @jstac @DrDrij
> 
> latest preview is available here
> 
> https://deploy-preview-121--sleepy-aryabhata-461b8b.netlify.app/lectures/#filter=*

@jstac I feel the `This website` text is redundant in this context -- perhaps they can all be "A set of lectures on ..."

jstac commented:
+1 regarding the change to "This webite..." Good catch.

I think we will have to do a manual ordering of the lectures in each section.  For example, in

![image](https://github.com/QuantEcon/website/assets/3887684/d43d7a08-2ba0-42cb-b29e-e68df017c411)

the order should be reversed -- python programming first, intro lectures next, intermediate one last.

That's the order they should be read in...

mmcky commented:
> +1 regarding the change to "This webite..." Good catch.
> 
> I think we will have to do a manual ordering of the lectures in each section. For example, in
> 
> ![image](https://private-user-images.githubusercontent.com/3887684/313032572-d43d7a08-2ba0-42cb-b29e-e68df017c411.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA0NjQwNTMsIm5iZiI6MTcxMDQ2Mzc1MywicGF0aCI6Ii8zODg3Njg0LzMxMzAzMjU3Mi1kNDNkN2EwOC0yYmEwLTQyY2ItYjI5ZS1lNjhkZjAxN2M0MTEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMxNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMTVUMDA0OTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjYwMmI1ZTZlMjZiZDQxZWMwN2Q4YWQxZTAwYmQ4NDE4NTU4ZjgxYmY4YTRmMmE4MWJjNzJiODI5MzNiMDNkYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.8ted2WtDWAXJ3sVXdCulp9YfdhdpwWFGFv8k49AEmRk)
> 
> the order should be reversed -- python programming first, intro lectures next, intermediate one last.
> 
> That's the order they should be read in...

+1 on the manual ordering -- I think that is a good idea. 

@DrDrij -- should that be set within groups or by the `order:` setting in the md files? 

In the topics section @jstac do you think we can set an order (for reading) or just an order of interest?

DrDrij commented:
Ahh yeah sorry for łka i've been looking at it for so long I forgot it was just all lorem ipsum. Thanks @mmcky 

Ok so I will add another metadata name/key pair for ordering.  

We can then prioritise that and see how we feel about it. 

Sorting by topic, order...

I think that means order by name and difficulty becomes obsolete?

DrDrij commented:
Yes to everything you’ve mentioned.

On Fri, 15 Mar 2024 at 12:43 am, Matt McKay ***@***.***>
wrote:

> thanks @DrDrij <https://github.com/DrDrij> nice work!
>
> Just a few quick questions:
>
>    - The structure is looking good - can I help with by populating the
>    lecture descriptions so we can preview it with Tom?
>    - can the new: key use new: {{ true / false }}? I would find that more
>    intuitive then 1
>
> Future Idea:
>
>    - would it be possible to add a published: {{ date }} field embedded
>    somewhere on the sites themselves and show an updated tag for lectures
>    that have been updated within the last month (or so).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuantEcon/website/pull/121#issuecomment-1998654604>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABM5AXOA6VK2V5NLXYDNMT3YYIYYVAVCNFSM6AAAAABEFW6WQSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTSOJYGY2TINRQGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

mmcky commented:
@DrDrij I have updated the lecture series names. 

If you have time early this week can we meet up to finalise this so we can merge. Thanks.

mmcky commented:
@DrDrij with the `All` category I think it would be most intuitive to sort by `Group` and then by `Group Order` so that the order in `All` is the same as the Groups  -- it's just that they are all visible. What do you think?

mmcky commented:
from @jstac 

With all the additions there's now a few too many colors --- perhaps we should drop the "new" labels?

Rather than opening on "All", which is probably a bit bewildering, perhaps it should open on "Introductory" by default and "All" should be the last option.

Once we switch to this default, perhaps having a single column of boxes is going to be simpler and cleaner.

In "Introductory", the order should be "Python Programming...", "A First...",  --- and I suppose the intermediate one will be gone?

With "Programming", it should be "Python Programming...", the JAX lectures, and then the Julia lectures (which are perhaps falling a bit behind -- I haven't seen many updates).

With "Topics" and "Statistics" I think you can choose.  With "All" it should be the easy ones first.

DrDrij commented:
> @DrDrij I have updated the lecture series names. 
> 
> 
> 
> If you have time early this week can we meet up to finalise this so we can merge. Thanks. 

Defiantly- I will be home by my machine in roughly 24 hours - and push through all the feedback on Tuesday to our branch. How about we arrange a time to chat after that @mmcky

DrDrij commented:
Getting there. Initial layout shows all instead of just Introduction (perhaps feature not bug?). However, GitHub giving warning of code vulnerabilities. Will look into that tomorrow.

DrDrij commented:
Order is now preserved from my testing, had to make certain the value was treated as an integer for items over 9...

![image](https://github.com/QuantEcon/website/assets/5886045/2ebc2674-be17-41f7-93bb-6c73c4a04833)

mmcky commented:
thanks @DrDrij I will loop through and edit the metadata to remove some duplicates and misclassified lectures

mmcky commented on _lectures/lecture-jax.md:
```diff
@@ -3,8 +3,8 @@ name: Quantitative Economics with JAX
 description: A set of lectures on quantitative economic modeling using GPUs and Google JAX.
 link: https://jax.quantecon.org/
 image: project-python.png
-order: 08
```
@DrDrij i have updated the order based on groups. Is that correct?

mmcky commented:
@DrDrij update the metadata is not super easy. Maybe we can investigate a central yml file at some point.

mmcky commented on _lectures/lecture-datascience.md:
```diff
@@ -3,7 +3,8 @@ name: Introduction to Economic Modeling and Data Science
 image: project-datascience.png
-order: 6
-type: lecture
+order: 3, 2
```
@DrDrij i have used the following syntax for within group ordering. Is this correct?

mmcky commented:
@DrDrij doesn't look like `All` is in the same order as the groups (listed one after the other)

DrDrij commented:
- [x] All is in the same order as the groups
- [x] Group ordering
- [ ] Central YML file?
- [x] Dropped ski-related icons for difficulty, replaced with number
- [x] Adjust primary navigation order, add lectures
- [x] Add 6 new lectures to lecture page
- [x] Change layout to grid
- [x] Add filtering
- [x] Add new metadata field for topics

mmcky commented:
thanks @DrDrij for your work here. We are getting close. 

<img width="433" alt="Screenshot 2024-03-22 at 5 30 54 PM" src="https://github.com/QuantEcon/website/assets/8263752/1ba54315-daca-43c8-8017-a4cc5a296570">

I prefer the colour coding (for level of difficulty) over the number (as I get confused re: is it an order etc.) I think it is great to drop the `new` but a little bit of colour might be nice.

DrDrij commented:
> @DrDrij doesn't look like `All` is in the same order as the groups (listed one after the other)

Is this one fixed for you @mmcky ?

DrDrij commented:
Okie. @mmcky Switched back to the colour coding difficulty system. I think it's much nicer than having the digit there as you say. I think this is just about ready to go — however, I will need to figure out a clean way to order via topics.
`order: 1, 2`
Doesn't respect the the topic the order belongs to. We could have:

```
order: 4, 6
topic: Introductory, Topics
```

```
topic: Introductory 4, Topics 6
```

That is potentially the simplest. Otherwise we need to go into array such as:

```
topics:
- Introductory
  4
- Topics
  6
```

Any preference?

mmcky commented:
@DrDrij This one

```
order: 4, 6
topic: Introductory, Topics
```

is the one I've done so far. It wasn't easy to edit. 

This 

```
topic: Introductory 4, Topics 6
```

would also be fine. 

It would be easier if we could define the order in single file (`_sorting.yml`) but no worries if that is not possible.

mmcky commented:
@DrDrij let me know when you've settled on a convention and the sorting is working -- then I think we can merge this and iterate from there.

DrDrij commented:
Are we potentially overthinking this? It appears to me it's just an edge case or two lectures series that we could *push* to the front, and in 'all' group them by the filter order? Just a thought.

Anyway, disregarding that I am assuming the _sorted.yml will look something like this @mmcky:

```
Introductory
- A First Course in Quantitative Economics
- An introduction to quantitative economics using Python.
- An introductory set of lectures on economic dynamics.
- A series of lectures on programming, data science, and economics.
- A set of lectures on Python programming for economics and finance.

Programming
- Python Programming for Economics and Finance
- A set of lectures on Python programming for economics and finance.
- A set of lectures on quantitative economic modeling using GPUs and Google JAX.
- A set of lectures on quantitative economic modeling.

Topic

~~~~~~~~~~
```

In this case, mapby we should use something a bit less user–input-prone for the title of each lecture series? An increasing digit? Series of chars?

mmcky commented:
> Are we potentially overthinking this? It appears to me it's just an edge case or two lectures series that we could _push_ to the front, and in 'all' group them by the filter order? Just a thought.
> 
> Anyway, disregarding that I am assuming the _sorted.yml will look something like this @mmcky:
> 
> ```
> Introductory
> - A First Course in Quantitative Economics
> - An introduction to quantitative economics using Python.
> - An introductory set of lectures on economic dynamics.
> - A series of lectures on programming, data science, and economics.
> - A set of lectures on Python programming for economics and finance.
> 
> Programming
> - Python Programming for Economics and Finance
> - A set of lectures on Python programming for economics and finance.
> - A set of lectures on quantitative economic modeling using GPUs and Google JAX.
> - A set of lectures on quantitative economic modeling.
> 
> Topic
> 
> ~~~~~~~~~~
> ```
> 
> In this case, mapby we should use something a bit less user–input-prone for the title of each lecture series? An increasing digit? Series of chars?

@DrDrij yeah I definitely don't think we should sort by the titles. (we would want to specify the docname probably). 

I am happy to work with what we have if we can get the sorting to work with the two levels. (Sorted by Group then Order in the `all` window)

DrDrij commented:
Ok @mmcky I need just jump off this for a short while. I think the display is tidy enough to go out as a beta.

I will come back to this after another piece of work I am doing. I am 100% certain that I can get the ordering to match that which is present in the metadata of each lecture-series document rather than alphabetical ordering in the topic groups.

Please give me your honest opinion and I'd be personally satisfied to get it running the way we want. Also please let me know if the full functionality is urgent and required for a demonstration.

Cheers.

mmcky commented:
thanks @DrDrij can I go ahead with the plan re: metadata

```md
order: <number>
topic: <name> <number>, <name> <number>
```

DrDrij commented:
@mmcky I've pushed up the latest updates. Been starting at it a while so please do a quick test on your machine too :)

Hey @jstac should have captured all your suggestions bro.

![image](https://github.com/QuantEcon/website/assets/5886045/5468fb58-ff4e-4894-ae7d-9c28ef0a8afa)

jstac commented:
Looks beautiful.  Many thanks @DrDrij !

mmcky commented:
thanks @DrDrij -- looking good.  

The `Topic` filter doesn't appear to obey the order in my commit [abf6db1](https://github.com/QuantEcon/website/pull/121/commits/abf6db15e54a0d6547e3288f0a6e071f7cda9654)

```
lecture-tools-techniques.md:topic: Topics 1
lecture-stats.md:topic: Topics 2
lecture-python.md:topic: Topics 3
lecture-julia.md:topic: Programming 3, Topics 4
lecture-datascience.md:topic: Introductory 3, Topics 5
lecture-jax.md:topic: Programming 2, Topics 6
lecture-eqm.md:topic: Topics 7
lecture-markov.md:topic: Topics 8
lecture-dynamics.md:topic: Topics 9
lecture-dps.md:topic: Topics 10
lecture-dle.md:topic: Topics 11
lecture-advanced.md:topic: Topics 12
```

@jstac Just working on the final order

DrDrij commented:
Hey @mmcky - had some luck, following the login of the ordering. We not have a separate lecture-series yaml file for each topic it lives under. So sorting, filtering, and ordering is all working. The remaining item is to go through the original Liquid array of lecture series and remove items that contain the same "title". ie. we don't need multiple items in the series sorting list with the same title but differing values.

So yeah, hopefully its all sorting nicely on your end but note that under "All" we do have each series that lives under 2 topics appearing twice.

<img width="1822" alt="image" src="https://github.com/QuantEcon/website/assets/5886045/32cbe396-2010-4d50-a3d9-79e6f928e114">

mmcky commented:
thanks @DrDrij so each list has its own set of `yml` files such as `topics-{name}.yml`?

> The remaining item is to go through the original Liquid array of lecture series and remove items that contain the same "title". ie. we don't need multiple items in the series sorting list with the same title but differing values.

Is this causing the duplicates in the `All` list?

mmcky commented:
> thanks @DrDrij so each list has its own set of `yml` files such as `topics-{name}.yml`?
> 
> > The remaining item is to go through the original Liquid array of lecture series and remove items that contain the same "title". ie. we don't need multiple items in the series sorting list with the same title but differing values.
> 
> Is this causing the duplicates in the `All` list?

@DrDrij is there any way to drop the duplicates in the `All` view?

DrDrij commented:
Hey @mmcky I've been tinkering away trying lots of ways to get the sorting to work. I've done written out the lectures on paper under each tab and the orders seem to be working. Mind giving it a test please :)

DrDrij commented:
Testing today I see it is still getting confused. I am using an array in the front matter

`difficulty: 3

topic:
  - 
    Introductory: 0
    Programming: 0
    Topics: 8
    All: 10
---
`

Will do a bit more testing.

mmcky commented:
thanks @DrDrij let me know if we should just write out the html manually for now. :-)

DrDrij commented:
Thanks @mmcky for the suggestion. I will do that now. I'm on my way back to Canberra on the 7th and will update to an automated method.

DrDrij commented:
@mmcky Latest commit is now static HTML ordering for now.

mmcky commented:
thanks @DrDrij

DrDrij commented:
@mmcky Might have another crack at this..

mmcky commented:
@DrDrij is this good to go or should we discuss before making this improvement. From memory the last update was that we would just manually update the locations of each lecture on the page. Is that right?

mmcky commented:
@DrDrij I will leave this open but it won't get merged. We can incorporate elements from this PR in our `org-site review`

DrDrij commented:
@mmcky Thanks for your summaries, I think we covered everything and this can be closed? 

If we'd like filters or the lectures in grids lets start a new issue and discuss :D

mmcky commented:
thanks @DrDrij
